File: frontend/src/views/products/repositories/productRepository.ts
import ApiService from '@/services/ApiService'
import { Product, CreateProductDto, UpdateProductDto, GetProductsQueryDto, ProductsResponse } from '../types/product.types'

export class ProductRepository {
  static async getProducts(query: GetProductsQueryDto = {}): Promise<ProductsResponse> {
    return ApiService.fetchDataWithAxios<ProductsResponse>({
      url: '/products',
      method: 'get',
      params: query,
    })
  }

  static async getProduct(id: string): Promise<Product> {
    return ApiService.fetchDataWithAxios<Product>({
      url: `/products/${id}`,
      method: 'get',
    })
  }

  static async createProduct(data: CreateProductDto): Promise<Product> {
    return ApiService.fetchDataWithAxios<Product>({
      url: '/products',
      method: 'post',
      data,
    })
  }

  static async updateProduct(id: string, data: UpdateProductDto): Promise<Product> {
    return ApiService.fetchDataWithAxios<Product>({
      url: `/products/${id}`,
      method: 'patch',
      data,
    })
  }

  static async deleteProduct(id: string): Promise<void> {
    return ApiService.fetchDataWithAxios<void>({
      url: `/products/${id}`,
      method: 'delete',
    })
  }
}

File: choppi-app/choppiapp/lib/features/products/data/repositories/product_repository.dart
import '../models/product_models.dart';
import '../services/product_service.dart';

class ProductRepository {
  final ProductService _productService;

  ProductRepository(this._productService);

  Future<ProductsResponse> getProducts({String? query, int page = 1, int limit = 10}) async {
    return await _productService.getProducts(query: query, page: page, limit: limit);
  }

  Future<Product> getProduct(String id) async {
    return await _productService.getProduct(id);
  }

  Future<Product> createProduct(CreateProductRequest request) async {
    return await _productService.createProduct(request);
  }

  Future<void> addProductToStore(String storeId, AddProductToStoreRequest request) async {
    return await _productService.addProductToStore(storeId, request);
  }
}